kythe/typescript/kythe.ts

```diff
@@ -0,0 +1,206 @@
+}
+
+/**
+ * A Kythe ordinal edge has the form of
+ *   `${EdgeKind}.${number}`
+ * This is represented as a branded string that is incompatible with a string
+ * but can be compared to a string.
+ */
+export type OrdinalEdge = string&{
+  __ordinalBrand: 'ordinal';
+};
+export function makeOrdinalEdge(edge: EdgeKind, ordinal: number): OrdinalEdge {
+  const ordinalEdge = `${edge}.${ordinal}`;
+  return <OrdinalEdge>ordinalEdge;
+}
+
+/**
+ * Kythe node kinds. See
+ *   https://kythe.io/docs/schema/#_node_kinds
+ *   https://github.com/kythe/kythe/tree/master/kythe/data/schema_index.textproto#L64
+ */
+export enum NodeKind {
+  ABS = 'abs',
+  ABSVAR = 'absvar',
+  ANCHOR = 'anchor',
+  CONSTANT = 'constant',
+  DIAGNOSTIC = 'diagnostic',
+  DOC = 'doc',
+  FILE = 'file',
+  INTERFACE = 'interface',
+  FUNCTION = 'function',
+  LOOKUP = 'lookup',
+  MACRO = 'macro',
+  META = 'meta',
+  NAME = 'name',
+  PACKAGE = 'package',
+  PROCESS = 'process',
+  RECORD = 'record',
+  SUM = 'sum',
+  SYMBOL = 'symbol',
+  TALIAS = 'talias',
+  TAPP = 'tapp',
+  TBUILTIN = 'tbuiltin',
+  TNOMINAL = 'tnominal',
+  TSIGMA = 'tsigma',
+  VARIABLE = 'variable',
+  VCS = 'vcs',
+}
+
+/**
+ * Kythe fact names. See
+ *   https://github.com/kythe/kythe/tree/master/kythe/data/schema_index.textproto#L92
+ */
+export enum FactName {
+  BUILD_CONFIG = '/kythe/build/config',
+  CODE = '/kythe/code',
+  COMPLETE = '/kythe/complete',
+  CONTEXT_URL = '/kythe/context/url',
+  DETAILS = '/kythe/details',
+  DOC_URI = '/kythe/doc/uri',
+  LABEL = '/kythe/label',
+  LOC_END = '/kythe/loc/end',
+  LOC_START = '/kythe/loc/start',
+  MESSAGE = '/kythe/message',
+  NODE_KIND = '/kythe/node/kind',
+  PARAM_DEFAULT = '/kythe/param/default',
+  RULE_CLASS = '/kythe/ruleclass',
+  SNIPPET_END = '/kythe/snippet/end',
+  SNIPPET_START = '/kythe/snippet/start',
+  SUBKIND = '/kythe/subkind',
+  TEXT = '/kythe/text',
+  TEXT_ENCODING = '/kythe/text/encoding',
+  VISIBILITY = '/kythe/visibility',
+}
+
+/**
+ * Kythe fact subkinds. See
+ *   https://github.com/kythe/kythe/tree/master/kythe/data/schema_index.textproto#L115
+ */
+export enum Subkind {
+  CATEGORY = 'category',
+  CLASS = 'class',
+  CONSTRUCTOR = 'constructor',
+  DESTRUCTOR = 'destructor',
+  ENUM = 'enum',
+  ENUM_CLASS = 'enumClass',
+  FIELD = 'field',
+  IMPLICIT = 'implicit',
+  IMPORT = 'import',
+  INITIALIZER = 'initializer',
+  LOCAL = 'local',
+  LOCAL_PARAMETER = 'local/parameter',
+  METHOD = 'method',
+  NAMESPACE = 'namespace',
+  STRUCT = 'struct',
+  TYPE = 'type',
+  UNION = 'union',
+}
+
+/**
+ * An Entry in the Kythe schema is either a Fact or an Edge that describes at
+ * least one node.
+ */
+export interface Entry {
+  source: VName;
+  label: string;
+}
+
+/**
+ * A Fact is an Entry that also has a fact `value`.
+ */
+export interface Fact extends Entry {
+  value: string;
+}
+
+/**
+ * An Edge is an Entry that also has a `target` and an edge `kind`.
+ */
+export interface Edge extends Entry {
```

Comment:
Should these be used by indexer.ts too?

Reply:
Yes, I'm going to add that in a separate PR.